File: examples/runnable/07-saga/python/saga.py
"""Order saga: authorize payment -> reserve inventory -> ship, with compensation.

The Python equivalent of the Java OrderSagaWorkflow lab. Python has no built-in
Saga helper, so we manage compensations by hand with a list used as a stack: each
forward step appends its undo, and on failure we run them in reverse.
"""

from datetime import timedelta

from temporalio import activity, workflow
from temporalio.common import RetryPolicy

TASK_QUEUE = "orders"


# --- Forward steps ---


@activity.defn
async def authorize_payment(order_id: str) -> str:
    return f"payment-{order_id}"


@activity.defn
async def reserve_inventory(order_id: str) -> str:
    return f"reservation-{order_id}"


@activity.defn
async def ship(order_id: str) -> None:
    # Fail on demand so you can trigger compensation from the CLI.
    if "fail" in order_id.lower():
        raise RuntimeError("shipping label service failed")


# --- Compensating steps (should be idempotent: they may be retried) ---


@activity.defn
async def cancel_payment(payment_id: str) -> None:
    activity.logger.info("cancelled %s", payment_id)


@activity.defn
async def restore_inventory(reservation_id: str) -> None:
    activity.logger.info("restored %s", reservation_id)


@activity.defn
async def send_failure_notification(order_id: str, reason: str) -> None:
    activity.logger.info("order %s failed: %s", order_id, reason)


@workflow.defn
class OrderSagaWorkflow:
    @workflow.run
    async def process(self, order_id: str) -> str:
        # Bound the retries so a permanent failure exhausts quickly and the except
        # block runs the compensations. Without maximum_attempts the default policy
        # retries forever and the saga never reaches compensation.
        opts = dict(
            start_to_close_timeout=timedelta(seconds=30),
            retry_policy=RetryPolicy(
                initial_interval=timedelta(milliseconds=500), maximum_attempts=3
            ),
        )

        # Each entry is (activity_fn, arg) to run if we have to unwind.
        compensations: list = []
        try:
            payment_id = await workflow.execute_activity(authorize_payment, order_id, **opts)
            compensations.append((cancel_payment, payment_id))

            reservation_id = await workflow.execute_activity(reserve_inventory, order_id, **opts)
            compensations.append((restore_inventory, reservation_id))

            await workflow.execute_activity(ship, order_id, **opts)
            return "COMPLETED"
        except Exception as failure:
            # Unwind in reverse: most-recent step backwards.
            for compensate, arg in reversed(compensations):
                await workflow.execute_activity(compensate, arg, **opts)
            await workflow.execute_activity(
                send_failure_notification, args=[order_id, str(failure)], **opts
            )
            return "COMPENSATED"

File: examples/runnable/07-saga/go/saga.go
// Package main holds the order saga lab: authorize payment -> reserve inventory
// -> ship, with compensation. The Go equivalent of the Java OrderSagaWorkflow lab.
// Go has no built-in Saga helper; the idiomatic pattern is a slice of compensation
// closures that we run in reverse on error.
package main

import (
	"context"
	"errors"
	"strings"
	"time"

	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

const TaskQueue = "orders"

// --- Forward steps ---

func AuthorizePayment(ctx context.Context, orderID string) (string, error) {
	return "payment-" + orderID, nil
}

func ReserveInventory(ctx context.Context, orderID string) (string, error) {
	return "reservation-" + orderID, nil
}

func Ship(ctx context.Context, orderID string) error {
	// Fail on demand so you can trigger compensation from the CLI.
	if strings.Contains(strings.ToLower(orderID), "fail") {
		return errors.New("shipping label service failed")
	}
	return nil
}

// --- Compensating steps (should be idempotent: they may be retried) ---

func CancelPayment(ctx context.Context, paymentID string) error {
	activity.GetLogger(ctx).Info("cancelled " + paymentID)
	return nil
}

func RestoreInventory(ctx context.Context, reservationID string) error {
	activity.GetLogger(ctx).Info("restored " + reservationID)
	return nil
}

func SendFailureNotification(ctx context.Context, orderID, reason string) error {
	activity.GetLogger(ctx).Info("order " + orderID + " failed: " + reason)
	return nil
}

// OrderSagaWorkflow runs the forward steps and, on failure, unwinds in reverse.
func OrderSagaWorkflow(ctx workflow.Context, orderID string) (string, error) {
	// Bound the retries so a permanent failure exhausts quickly and we reach
	// compensation. Without MaximumAttempts the default policy retries forever and
	// the saga never reaches the compensation path.
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 30 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval: 500 * time.Millisecond,
			MaximumAttempts: 3,
		},
	})

	// Each compensation is a closure capturing the id it needs to undo.
	var compensations []func()
	compensate := func() {
		// Reverse order: last registered runs first.
		for i := len(compensations) - 1; i >= 0; i-- {
			compensations[i]()
		}
	}

	var paymentID string
	if err := workflow.ExecuteActivity(ctx, AuthorizePayment, orderID).Get(ctx, &paymentID); err != nil {
		return "FAILED", err
	}
	compensations = append(compensations, func() {
		_ = workflow.ExecuteActivity(ctx, CancelPayment, paymentID).Get(ctx, nil)
	})

	var reservationID string
	if err := workflow.ExecuteActivity(ctx, ReserveInventory, orderID).Get(ctx, &reservationID); err != nil {
		return compensateAndNotify(ctx, compensate, orderID, err)
	}
	compensations = append(compensations, func() {
		_ = workflow.ExecuteActivity(ctx, RestoreInventory, reservationID).Get(ctx, nil)
	})

	if err := workflow.ExecuteActivity(ctx, Ship, orderID).Get(ctx, nil); err != nil {
		return compensateAndNotify(ctx, compensate, orderID, err)
	}

	return "COMPLETED", nil
}

func compensateAndNotify(ctx workflow.Context, compensate func(), orderID string, cause error) (string, error) {
	compensate()
	_ = workflow.ExecuteActivity(ctx, SendFailureNotification, orderID, cause.Error()).Get(ctx, nil)
	return "COMPENSATED", nil
}
